### PR TITLE
Fix #5653: improve UserPermissions - password mutations

### LINKED ISSUE #5653: Missing Change Password in user permissions?
**Describe the bug**
🤞 I'm not missing something but it looks like the current `/change-password` REST and/or the
`changePassword` GraphQL mutation actually is handling the **reset password** logic. 

To fix this I'm proposing two things:
1. rename current `change-password` logic to `reset-password`
2. implement a new `change-password` mutation, for which I have a proposed PR, which I will reference in the comments shortly.

**Expected behavior**
For the global self-isolation to end soon.

**Proposed solution**
**Follow the progress:** https://github.com/strapi/strapi/pull/5655

**Inpired by**
The awesome PR https://github.com/strapi/strapi/pull/5580/ by @roelbeerens

### PR DESCRIPTION
**Pull request by ScottAgirs**
### The Problem
Current reset-password endpoints are pointing to "changePassword" Auth controller which is misleading and does not actually do the **change password** action in the traditional sense, it rather does what the endpoint says - resets user's password.

### Intent of this PR
To conform and improve reset password naming and yield way for potential future implementation of an actual **change password** functionality, while at the same time making it easier and less confusing for the user to navigate and orient oneself in the password mutations.

~To enable users to be able to change their password with their current password (without the need to generate `code`).~
~Currently, unless badly overlooking something, it's only possible by resetting the password via forgot-password.~

#### Added
- documentation for GraphQL reset password mutation
~- `change-password` mutation (replaces the previous `change-password` - which if this is merged becomes `reset-password`)~
~- GraphQL `changePassword` mutation (replaces the previous `changePassword` - which if this is merged becomes `resetPassword`)~

#### Refactored
~- **[BREAKING]** rename `change-password` to `reset-password`~
- renamed the Auth controller from `changePassword` to `resetPassword`
- renamed response payload in GraphQL from `ForgotPassword` to `UserPersmissionsPasswordPayload`

### CODE REVIEW

**Comment by alexandrebodin** [comment]
Can you explain the intent of the PR in the description plz ?

**Comment by ScottAgirs** [comment]
@alexandrebodin done, hope this makes sense

**Comment by roelbeerens** [comment]
@ScottAgirs I like the idea of a changePassword feature! Also, thank you for providing the documentation (I kinda forgot about those). 

The only thing that worries me right now on the changePassword mutation is, that it can only by used with providing a Bearer token with the request. The general idea is that the user should be logged in to change it's password, of course. However, I do think the 'endpoint' / controller should also be available on the 'public' side (if you want to when setting the permissions). I think the mutation should have a 'code' variable/parameter available to resolve this. This however, does make the mutation pretty much redundant and a copy of the current 'changePassword' mutation, except for the current password confirmation. That would be the main difference.

Here's what I think that's missing:

- Leave the renaming as it is, I think it's more logical this way.
- Add a 'code' parameter to the changePassword mutation (the code is a JWT token anyway and should contain the current user ID)
- Instead of using the 'state.user' in the controller, decode the code and use that user information.

With this, the endpoint/controller can be used both as a 'public' user, as a 'authenticated' user.

If I overlooked something, please let me know, I didn't fully dug through your PR ;)

Great rename on the ForgotPassword payload btw!

**Comment by alexandrebodin** [comment]
Hi, for security reasons, I really don't think a call to an api should be able to change a user password. I would rather only have the reset password workflow. Same for the administration of users. I think we should change this to only be able to send a reset link to the user.

**Comment by roelbeerens** [comment]
@alexandrebodin You've got a point there.

**Comment by ScottAgirs** [comment]
@roelbeerens thanks for the feedback! Was hoping you'll be ok with the ForgotPassword payload rename, haha! The token is a nice idea!

@alexandrebodin I understand, but changing password is quite a basic and important ability to have for the users, no? Or would you guys say that changing the password by the current password is a thing of the past?

**Comment by ScottAgirs** [comment]
PS. as an extra security layer we could add email notification about the action - in case we do decide to move forward with this

**Comment by alexandrebodin** [comment]
I think this topics needs more thinking to be put into it before take a decision about this. I would be ok with an api call that would trigger a reset password flow for now only.

**Comment by ScottAgirs** [comment]
> I think this topics needs more thinking to be put into it before take a decision about this. I would be ok with an api call that would trigger a reset password flow for now only. This would be equivalent to the forgot password basically

That's fair. I'll adjust the PR, so that the edits are only related to the reset password flow

**Comment by ScottAgirs** [comment]
@alexandrebodin I have cleaned up, updated the intent and defined the problem in the description. Please take a look and let me know what you think.
@roelbeerens perhaps you could chime in too?

**Comment by alexandrebodin** [comment]
This looks good to me, this will however be a breaking change, I'm flagging it as so and will merge it after our next release

**Comment by alexandrebodin** [comment]
Hi @ScottAgirs We have made some updates to files you ahve changed recently would you mine resolving the conflicts so we can move this PR into the next release ?

**Comment by ScottAgirs** [comment]
> Hi @ScottAgirs We have made some updates to files you ahve changed recently would you mine resolving the conflicts so we can move this PR into the next release ?

Hi @alexandrebodin thanks for the update, just pushed

File: packages/strapi-plugin-users-permissions/config/schema.graphql.js [hunk 1]
```diff
@@ -53,18 +53,18 @@ module.exports = {
...
   `,
   query: `
     me: UsersPermissionsMe
   `,
   mutation: `
     login(input: UsersPermissionsLoginInput!): UsersPermissionsLoginPayload!
-    register(input: UsersPermissionsRegisterInput!): UsersPermissionsLoginPayload!
-    forgotPassword(email: String!): ForgotPassword
-    changePassword(password: String!, passwordConfirmation: String!, code: String!): UsersPermissionsLoginPayload
+    register(input: UserInput!): UsersPermissionsLoginPayload!
```
Comment (alexandrebodin) [inline packages/strapi-plugin-users-permissions/config/schema.graphql.js:65]: Small mistake when you fixed the conflicts :) We should keep the UsersPermissionsRegisterInput :)

Comment (alexandrebodin) [reply packages/strapi-plugin-users-permissions/config/schema.graphql.js:65]: Hi @ScottAgirs  will you be able to make the change ?

Comment (ScottAgirs) [reply packages/strapi-plugin-users-permissions/config/schema.graphql.js]: @alexandrebodin hey Alex, thanks for the follow-up, my apologies, I must have missed the notification about your review. 

Just pushed 🚀

File: packages/strapi-plugin-users-permissions/config/schema.graphql.js [hunk 2]
```diff
@@ -53,18 +53,18 @@ module.exports = {
...
   `,
   query: `
     me: UsersPermissionsMe
   `,
   mutation: `
     login(input: UsersPermissionsLoginInput!): UsersPermissionsLoginPayload!
-    register(input: UsersPermissionsRegisterInput!): UsersPermissionsLoginPayload!
-    forgotPassword(email: String!): ForgotPassword
-    changePassword(password: String!, passwordConfirmation: String!, code: String!): UsersPermissionsLoginPayload
+    register(input: UserInput!): UsersPermissionsRegisterInput!
```
Comment (alexandrebodin) [inline packages/strapi-plugin-users-permissions/config/schema.graphql.js:65]: You forgot the input too :D `UsersPermissionsRegisterInput `

Comment (ScottAgirs) [reply packages/strapi-plugin-users-permissions/config/schema.graphql.js]: ahh lol ebed726d7

File: packages/strapi-plugin-users-permissions/config/schema.graphql.js [hunk 3]
```diff
@@ -57,8 +57,8 @@ module.exports = {
   mutation: `
     login(input: UsersPermissionsLoginInput!): UsersPermissionsLoginPayload!
     register(input: UserInput!): UsersPermissionsLoginPayload!
-    forgotPassword(email: String!): ForgotPassword
-    changePassword(password: String!, passwordConfirmation: String!, code: String!): UsersPermissionsLoginPayload
+    forgotPassword(email: String!): UserPersmissionsPasswordPayload
```
Comment (sanojsilva) [inline packages/strapi-plugin-users-permissions/config/schema.graphql.js:66]: Hey there's a spelling mistake here User**Persmissions**PasswordPayload** it should be User**Permissions**PasswordPayload

Comment (alexandrebodin) [reply packages/strapi-plugin-users-permissions/config/schema.graphql.js:66]: nice catch, at least the code works :) Mind opening a PR ?

Comment (ScottAgirs) [reply packages/strapi-plugin-users-permissions/config/schema.graphql.js:66]: nice catch, Sanoj, hope u don't mind I had it open, so I just pushed the patch :) https://github.com/strapi/strapi/pull/6390

**Pull request merged**